.github/workflows/windows.yml

```diff
@@ -46,42 +46,163 @@ env:
+
+          # Generate a unique id that allows the workload job to find the runner
+          # we are reserving for it (via runner labels), and allows the timeout
+          # job to find the workload job run (via the job’s friendly name), even
+          # if there are multiple instances in the workflow call tree.
+          random32=$(LC_ALL=C < /dev/urandom tr -dC 0-9a-f 2> /dev/null | head -c 8)
+          unique_id=${{ github.repository }}/${{ github.run_id }}/$random32
+          echo "unique_id=$unique_id" | tee -a $GITHUB_OUTPUT
+
+          # RUNNER_API_TOKEN secret will be unavailable in forks.
+          if [ -z "$GH_TOKEN" ]; then
+            fall_back_to_github_hosted
+          fi
+
+          runners=$(mktemp)
+          gh api "$self_hosted_runner_scope" > $runners
+
+          # Find a runner that is online, not busy, and not already reserved for
+          # any job (label prefix “reserved-for:”).
+          runner_id=$(mktemp)
+          if ! < $runners > $runner_id jq \
+            --arg self_hosted_image_label "$self_hosted_image_label" -e '
+              .runners
+              | map(select(.status == "online" and .busy == false))
+              | map(select([.labels[].name] | index($self_hosted_image_label) | not | not))
+              | map(select([.labels[].name] | map(startswith("reserved-for:")) | index(true) | not))
+              | first | .id'; then
+            fall_back_to_github_hosted
+          fi
+
+          # Reserve that runner by adding a label containing the unique id.
+          # Job concurrency ensures that runners never get assigned twice.
+          reserved_since=$(date +\%s)
+          gh api "$self_hosted_runner_scope/$(cat $runner_id)/labels" \
+            -f "labels[]=reserved-for:$unique_id" \
+            -f "labels[]=reserved-since:$reserved_since" \
+            --method POST --silent
+          echo "selected_runner_label=reserved-for:$unique_id" | tee -a $GITHUB_OUTPUT
+          echo 'is_self_hosted=true' | tee -a $GITHUB_OUTPUT
+
+  # In the unlikely event a self-hosted runner was selected and reserved but it
+  # goes down before the workload starts, cancel the workflow run.
+  runner-timeout:
+    needs:
+      - runner-select
+    if: ${{ fromJSON(needs.runner-select.outputs.is-self-hosted) }}
+    name: Runner timeout
+    runs-on: ubuntu-latest
+    steps:
+      - name: Wait a bit
+        run: sleep 30
+
+      - name: Cancel if workload job is still queued
+        run: |
+          run_url=/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}
+          export GH_TOKEN=${{ secrets.GITHUB_TOKEN }}
+
+          if [ "$(gh api "$run_url/jobs" \
+            | jq -er --arg id '${{ needs.runner-select.outputs.unique-id }}' \
+              '.jobs[] | select(.name | contains("[" + $id + "]")) | .status'
+          )" = queued ]; then
+            echo "Timeout waiting for runner assignment! Cancelling workflow run"
+            gh api "$run_url/cancel" --method POST
```

Comment:
we should fail job here to make it more obvious what is wrong with simple `exit 1` and maybe clear reservation labels. Actually are reservation labels actually cleared anywhere? If not we should add job that runs always when self hosted is selected (no matter the result of timeout or build job) that will clear up labels.

Reply:
done in 4b3e3d782b5e0fcf0f36c4d2e477bd0e144c7749, though `reserved-for:` labels are already cleared by the monitor ([367f17174c872](https://github.com/servo/ci-runners/commit/367f17174c8724f6f866315cf1ca38341b206b2e)). i think that’s good enough, because we don’t really want to touch runner labels outside of job concurrency, and the runner might fail to assign or run the job for a more interesting reason that has side effects on the runner.

Reply:
Oh, I didn't know we also run monitor for runners, interesting. Are labels also cleared if job is done, but timeout is not hit yet? What if runner is running longer than `SERVO_CI_MONITOR_RESERVE_TIMEOUT`, will this unreserved runner even if it's still running (being reserved)?

Reply:
the only time we clear labels is when a runner gets unregistered, either automatically due to its job finishing, or if the monitor notices the runner was [reserved but unassigned](https://github.com/servo/ci-runners/blob/cdc1481f9b4f8f7305c3aba976f1c51bc4c54942/monitor/src/runner.rs#L184-L192) for SERVO_CI_MONITOR_RESERVE_TIMEOUT. if the runner gets assigned, that timeout no longer applies.

Reply:
ok